aws/resource_aws_glue_catalog_database.go

```diff
@@ -0,0 +1,180 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/glue"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsGlueCatalogDatabase() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsGlueCatalogDatabaseCreate,
+		Read:   resourceAwsGlueCatalogDatabaseRead,
+		Update: resourceAwsGlueCatalogDatabaseUpdate,
+		Delete: resourceAwsGlueCatalogDatabaseDelete,
+		Exists: resourceAwsGlueCatalogDatabaseExists,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"catalog_id": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"location_uri": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"parameters": {
+				Type:     schema.TypeMap,
+				Elem:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
```

Comment:
What is the reason for setting the above 3 fields as `Computed`? I tried removing the flag and all acceptance tests passed which makes me think there is no reason, but maybe I'm wrong and the behaviour just isn't covered  in tests?

Reply:
@radeksimko the motivation is because of the above discussion re computed flags with @bflad 

bflad:
>  In general, we want to know if something occurred outside Terraform as well (e.g. a manual update in the UI). I believe the intent of Computed: true for the other optional attributes is to catch these scenarios. If you look around at other resource schemas, you should see that the two are often paired together.

I don't know the functional requirements here, and testing the functionality describe i the quote here to import would require a step in between each `resource.TestStep` which goes into AWS and modifies the AWS resource.

I'm happy to do this, but I have not seen any prior art as to how one would go about this. But also, I would like to have some sort of idea about what is expected in terms of functionality so I don't end up bouncing back and forth between two different opinions of Reviewers :-)

I'm not sure who is closer to the project between yourself @radeksimko and @bflad :-)

Reply:
I will ask @bflad about this when he wakes up.

Reply:
Just confirmed with @bflad privately that this was just a misunderstanding and we mutually agreed it should be removed from the schema here.

Sorry for any confusion caused by this!

After that I believe we're ready to merge it 🎉

Reply:
For my own understanding, aside from obvious things (like instance ids i-12345678, etc), when should the `Computed:` tag be used?

Reply:
`Computed` w/ `Optional` is typically used when the API automatically sets unpredictable default value or when the value is generally not static and depends on other attributes.

It is an expectation of Terraform users to see difference in `plan` if something has changed outside of Terraform (e.g. someone manually changed description through AWS Console or AWS CLI). User has some options to ignore certain changes if they wish (via `lifecycle` block), but overall this is a desired behaviour.

The downside of marking something `Computed` is that it will ignore such changes, but sometimes it's necessary. We always prefer static `Default` where possible.

`Computed` only is used for any IDs or anything not defined in the config and returned by the API.

Reply:
So, there is a case above https://github.com/terraform-providers/terraform-provider-aws/pull/2175#discussion_r157543208 where I have a value (`catalog_id`), which at the moment is *only* able to be set as an AWS AccountID (not necessarily the same AccountID as TF is running as). If the `catalog_id` is omitted, AWS will fill it in.

I assume this is because AWS will eventually have the ability for multiple catalogues in a single account, a globally unique namespace for public catalogues, or something such. This then is a case where the user could provide `catalog_id`, but if they don't AWS will compute it.

This variable is currently `Computed: true` and given what you said above it seems like it is an ID, but... it could be a user set value, so it feels like a grey area, but... if it changes, the resource needs to be re-built, so it should never turn up as a diff in an update operation.

But, I would be correct in saying that this `catalog_id` is a valid use case for `Computed: true`?

Reply:
That's right, `catalog_id` is a valid use case. It's kind of a special case I didn't mention in my previous explanation.